src/fsharp/block.fs

```diff
@@ -0,0 +1,222 @@
+        if arr1.Length <> arr2.Length then
+            invalidOp "Block lengths do not match."
+
+        match arr1.Length with
+        | 0 -> true
+        | 1 -> predicate arr1.[0] arr2.[0]
+        | n ->
+            let mutable result = true
+            let mutable i = 0
+            while result && i < n do
+                result <- predicate arr1.[i] arr2.[i]
+                i <- i + 1
+            result
+
+    let tryFind predicate (arr: block<'T>) =
+        match arr.Length with
+        | 0 -> None
+        | 1 -> if predicate arr.[0] then Some arr.[0] else None
+        | n ->
+            let mutable result = None
+            let mutable i = 0
+            while result.IsNone && i < n do
+                if predicate arr.[i] then
+                    result <- Some arr.[i]
+                i <- i + 1
+            result
+
+    let tryFindIndex predicate (arr: block<'T>) =
+        match arr.Length with
+        | 0 -> None
+        | 1 -> if predicate arr.[0] then Some 0 else None
+        | n ->
+            let mutable result = None
+            let mutable i = 0
+            while result.IsNone && i < n do
+                if predicate arr.[i] then
+                    result <- Some i
+                i <- i + 1
+            result
+
+    let tryPick chooser (arr: block<'T>) =
+        match arr.Length with
+        | 0 -> None
+        | 1 -> chooser arr.[0]
+        | n ->
+            let mutable result = None
+            let mutable i = 0
+            while result.IsNone && i < n do
+                result <- chooser arr.[i]
+                i <- i + 1
+            result
+
+    let ofSeq (xs: 'T seq) =
+        ImmutableArray.CreateRange(xs)
+
+    let append (arr1: block<'T1>) (arr2: block<'T1>) : block<_> =
+        arr1.AddRange(arr2)
+
+    let createOne (item: 'T) : block<_> =
+        ImmutableArray.Create(item)
+
+    let filter predicate (arr: block<'T>) : block<'T> =
+        let builder = ImmutableArray.CreateBuilder(arr.Length)
+        for i = 0 to arr.Length - 1 do
+            if predicate arr.[i] then
+                builder.Add(arr.[i])
+        builder.ToImmutable()
+
+    let exists predicate (arr: block<'T>) =
+        let n = arr.Length
+        let mutable result = false
+        let mutable i = 0
+        while not result && i < n do
+            if predicate arr.[i] then
+                result <- true
+            i <- i + 1
+        result
+
+    let choose (chooser: 'T -> 'U option) (arr: block<'T>) : block<'U> =
+        let builder = ImmutableArray.CreateBuilder(arr.Length)
+        for i = 0 to arr.Length - 1 do
+            let result = chooser arr.[i]
+            if result.IsSome then
+                builder.Add(result.Value)
+        builder.ToImmutable()
+
+    let isEmpty (arr: block<_>) = arr.Length = 0
+
+    let fold folder state (arr: block<_>) =
+        let f = OptimizedClosures.FSharpFunc<_, _, _>.Adapt(folder)
```

Comment:
You probably want this on a few of the other functions too, it still really makes a big difference with little downside (2,5x perf for simple folders/choosers/etc).

See https://github.com/dotnet/fsharp/issues/6499#issuecomment-869195827